RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+backends. This is problematic, so we also generate a separate dispatch
+key handler which will automatically handle shape checking and other
+concerns for backends that are not concerned with performance:
+
+```
+// Name TBD; for this example we will call it DispatchKey::Common, as the
+// functionality here is common to all backends. This is an alias key
+// that resolves CommonXLA/CommonMSNPU/... in the same way as Autograd.
+
+Tensor upsample_nearest1d_common(const Tensor& self, IntArrayRef output_size, optional<double> scales) {
+  // TODO: RecordFunction could be added here, if desired
+  DeviceGuard g(self.device());
+  TensorMeta expected = meta::upsample_nearest1d(self, output_size, scales);
+  Tensor result = at::empty(expected.sizes(), expected.options());
+  ExcludeDispatchKeyGuard g2(DispatchKey::Common);
+  // Notice that upsample_nearest1d is ignored here. It may be a good idea
+  // to skip this implementation, or use a slightly different variant, if
+  // a backend explicitly registered upsample_nearest1d
+  at::upsample_nearest1d_out(expected, result, self, output_size, scales);
+  return result;
+}
+
+TORCH_LIBRARY_IMPL(aten, Common, m) {
+  m.impl("upsample_nearest1d", upsample_nearest1d_common);
+}
+```
+
+A backend that would like to fuse this boilerplate into their kernel for
+performance reasons can simply override the Common entry with
+fallthrough:
+
+```
+TORCH_LIBRARY_IMPL(aten, CommonXLA, m) {
+  m.impl("upsample_nearest1d", CppFunction::makeFallthrough);
+}
+```
+
+Finally, structured kernels also implicitly register an implementation
+for the Meta key, which is the API for dry running shape/dtype
+calculations without any kernels involved:
+
+```
+Tensor upsample_nearest1d_meta(const Tensor& self, IntArrayRef output_size, optional<double> scales) {
+  TensorMeta expected = meta::upsample_nearest1d(self, output_size, scales);
+  return native::empty_meta(expected.sizes(), expected.options());
+}
+
+TORCH_LIBRARY_IMPL(aten, Meta, m) {
+  m.impl("upsample_nearest1d", upsample_nearest1d_meta);
+}
+```
+
+## Discussion
+
+* Why not just use the Common dispatch key for everything?
+
+    * Performance reasons. Introducing the common key would induce an
+      extra redispatch, which at time of writing would give up quite a
+      bit of performance due to dispatch overhead, for no particularly
+      good reason.
+
+* Will this increase binary size?
+
+    * Not necessarily. By-in-large, all of the boilerplate code
+      generated here is already duplicated at the source level today
+      (for example, many operators have shape checks factored into
+      header files that are then shared between two operators).
+
+* I don’t like codegen.
+
+    * An earlier version of this proposal had the boilerplate
+      generated using C++ templates rather than codegen. However, we
+      think the formulation in this proposal is superior under the
+      constraint that mobile selective build must keep working, as we
```

Comment:
I don't understand the connection to mobile selective build yet

Reply:
We can talk about this in the meeting. It's constraint solving from the problem "mobile requires registrations to be a separate compilation unit from kernels"